### DISCUSSION TITLE
Path based bin or bin cut outs

### DISCUSSION THREAD

**Discussion by unteins** [kennetek/gridfinity-rebuilt-openscad#10] [Ideas]
Another enhancement that would be really neat is the ability to cut an arbitrary shape into a bin section.

Use case here is to create bins with cutouts to more closely hold a tool or specific shape. 

At a basic level think about cutting a circle into the bin section to hold a battery or a glue stick. But rather than just cutting a circle straight down to some depth, you could cut any arbitrary path to the depth to match the contours.

It would be even nicer if the top or bottom edge could have a radius to make a smoother pocket for the tool.

**Comment by kennetek** [comment]
You can do this already with the script. See documentation for the `cut_move` function for a full explanation. Basically, if you make the cut profile with OpenSCAD's 2D engine, centered about the origin, and then call something like `linear_extrude` to whatever depth is desired, and finally a `cut_move` to the desired location. There is an example in the script that does this with a cylinder. The pseudocode would look something like this:
```
gridfinityCustom() {
    // whatever other desired cuts
    cut_move(...) 
    linear_extrude(2*desired_depth, center=true)
    // 2D profile
}
```
This was the reason for exposing the `cut_move` function, to allow people to easily locate and cut custom shapes, as usually it was only used internally by the compartment spacing logic. 

As for smoothing the top or bottom edge, fillets are difficult to do in OpenSCAD, especially for arbitrary profiles. I can imagine a way to add a chamfer to either edge, and a fillet to the bottom (but not the top) edge. 

A module that does the extrusion and fillet/chamfer automatically on a given profile may be useful.

**Reply by unteins** [reply]
I'll go look at CutMove in more detail - didn't realize it had that much power, but nice!

**Comment by kennetek** [comment]
I tried to add something that added fillets or chamfers to cutouts, but it would only work for fully convex profiles. OpenSCAD just doesn't like fillets and chamfers, unfortunately.

**Reply by unteins** [reply]
Thanks for trying.

For what I was doing I could probably cheat by using a rounded cube as the cutting object, leaving behind filleted internal edges.

**Reply by kennetek** [reply]
That's exactly how I was going about it. But then, that rounds the side edges of the cutout as well, not just the bottom (now you have a rounded square cutout, instead of a square, for instance). So it ends up only working for certain shapes, like cylinders. Same with chamfers, it involved using a hull() command, which then ruined the original profile shape. So there is no arbitrary solution, but certain shapes can be done.

**Reply by unteins** [reply]
I found a rounded cube module that allowed you to round only certain edges of the cube - so you could still have a perfectly square top cutout with fillets at the bottom.

You can see this in my screwdriver bin where the corners of the lowered top are round, but there's no fillet between the walls and the horizontal part.
